File: Jonghyeok Kim/2022-10-23-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        counter = 0
        
        def dfs(i, j):
            if i == len(grid) or j == len(grid[0]) or i == -1 or j == -1:
                return
            if grid[i][j] == "0":
                return

            
            grid[i][j] = "0"
            
            dfs(i-1,j)
            dfs(i+1,j)
            dfs(i,j-1)
            dfs(i,j+1)
            
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1":
                    counter += 1
                    dfs(i,j)
                    
        return counter

File: HyeKyungKo/2022-10-23-Number of Islands.java
//2022.10.23
//Time Complexity: O(MxN)
//Space Complexity: O(MxN)
class Solution{    
    
    public int numIslands(char[][] grid){
        
        if(grid == null || grid.length == 0 || grid[0].length == 0){
            return 0;
        }
        
        int rowSize = grid.length;
        int colSize = grid[0].length;
        
        int totalNumber = 0;
        
        for(int i = 0; i < rowSize; i++){
            for(int j = 0; j < colSize; j++){
                if(grid[i][j] == '1'){
                    visitIslands(grid, i, j);                                 
                    totalNumber++;
                }

            }
        }
        
        return totalNumber;
    }
    
    private void visitIslands(char[][] grid, int row, int col){
        
        int rowSize = grid.length;
        int colSize = grid[0].length;
        
        if(row < 0 || row >= rowSize || col < 0 || col >= colSize){
            return;
        }
        
        if(grid[row][col] == '0'){ // ocean
            return;
        }
        
        grid[row][col] = '0'; //check as visited
        
        visitIslands(grid, row-1, col); //left
        visitIslands(grid, row, col-1); // up
        visitIslands(grid, row+1, col); //right
        visitIslands(grid, row, col+1); //down        
        
    }
}